parachain-system/src/lib.rs

```diff
@@ -1351,4 +1524,300 @@ mod tests {
 				},
 			);
 	}
+
+	#[test]
+	fn receive_dmp() {
+		BlockTests::new()
+			.with_validation_data(
+				|_, relay_block_num, validation_data| match relay_block_num {
+					1 => {
+						validation_data.dmq_mqc_head = hex![
+							"0f62484bb072376abd2a05bcac252b023e4564242b3487c978a860d6312af7cd"
+						]
+						.into();
+					}
+					_ => unreachable!(),
+				},
+			)
+			.with_inherent_data(|_, relay_block_num, data| match relay_block_num {
+				1 => {
+					data.downward_messages.push(InboundDownwardMessage {
+						sent_at: 1,
+						msg: b"down".to_vec(),
+					});
+				}
+				_ => unreachable!(),
+			})
+			.add(1, || {
+				HANDLED_DOWNWARD_MESSAGES.with(|m| {
+					let mut m = m.borrow_mut();
+					assert_eq!(
+						&*m,
+						&[InboundDownwardMessage {
+							sent_at: 1,
+							msg: b"down".to_vec(),
+						}]
+					);
+					m.clear();
+				});
+			});
+	}
+
+	#[test]
+	fn receive_hrmp() {
+		BlockTests::new()
+			.with_validation_data(
+				|_, relay_block_num, validation_data| match relay_block_num {
+					1 => {
+						// 200 - doesn't exist yet
+						// 300 - one new message
+						validation_data.hrmp_mqc_heads.push((
+							ParaId::from(300),
+							hex![
+								"d8d73e427fbcb01793aae939418c28af9f570da4f3d8430076a58844397c78a5"
+							]
```

Comment:
Maybe we should have a function for calculating this hash. We could use it in the production code and also in tests to not hardcode these values.

Reply:
Yeah, I considered that.

I deliberately chose this hardcoded snapshot like approach. My reasoning was that if you inadvertently changed the function, it will likely pass the tests, whereas it should not. If you do want to change the function, that will involve changing the hashes which is admittedly annoying, but still not that bad and unlikely to happen.

Reply:
On the second thought, we can extract the function (or functions, because there are two actually) and then test that it returns expected results.

Reply:
https://github.com/paritytech/cumulus/pull/308/commits/388168e300246237f3cfecc9f1a8740e208128dd